### ISSUE TITLE
Can not assign a non-sendable but sending value into inout sending value

### ISSUE THREAD

**Issue by muukii** [swiftlang/swift#77199]
### Reproduction

```swift
import Synchronization

final class MutexExample {
  
  final class NS {
    
  }
  
  let ns_mutex = Mutex<NS>.init(.init())
      
  func up() {
    
    let newNS = NS()
    
    ns_mutex.withLock { value in
      // error here
      value = newNS
    }
     
  }
  
}
```

### Expected behavior

Should not be able to compile?

### Environment

- Xcode 16.1 RC
- Swift6 mode

**Comment by jamieQ** [comment]
the following workaround appears to subvert the compiler error at the moment (at least with the recent nightly builds):

```swift
public func up() {
    let newNS = NS()
    let workAround = { newNS }

    ns_mutex.withLock { value in
        value = workAround()
    }
}
```

this at least offers a means to avoid calling the constructor within the mutex.

**Comment by muukii** [comment]
Thank you for sharing the workaround.
I updated the description adding the actual error what they saying

> 'inout sending' parameter 'value' cannot be task-isolated at end of function

What is meaning?
I'm still catching up on the concurrency model but is the closure of withLock not a sendable closure, so it's not task-isolated?

**Comment by jamieQ** [comment]
> What is meaning?
> I'm still catching up on the concurrency model but is the closure of withLock not a sendable closure, so it's not task-isolated?

yes, the `Mutex`'s `withLock` closure is not sendable, but i think the `sending` attributes on the closure parameter and return value are intended to ensure that the protected value cannot change isolation domains within the closure. my interpretation of the error in this case is that the `value` protected by the lock is non-sendable and in a 'disconnected' isolation region and the local variable `newNS` is for some reason being inferred as task-isolated, and this conflicts with the initial isolation of the value.

admittedly, i'm not an expert on how this is supposed to work, but it seems like there might be a bug here, and the error messaging in this case could probably be improved. if interested, here are some portions of the relevant Swift evolution proposals that may help shed more light on the issue:

1. '[Task-isolated regions](https://github.com/swiftlang/swift-evolution/blob/93575ecac1d292970a389a144d8fc9187d94259c/proposals/0414-region-based-isolation.md#task-isolated-regions)'
2. [inout behavior with sending parameters](https://github.com/swiftlang/swift-evolution/blob/93575ecac1d292970a389a144d8fc9187d94259c/proposals/0430-transferring-parameters-and-results.md#inout-sending-parameters)
3. [Mutex interactions with Swift concurrency](https://github.com/swiftlang/swift-evolution/blob/93575ecac1d292970a389a144d8fc9187d94259c/proposals/0433-mutex.md#interactions-with-swift-concurrency) (note that the terminology is somewhat outdated as 'transferring' was renamed to 'sending')
